### PR TITLE
MacOS Apple Silicon (arm64) cross-build support

### PR DESCRIPTION
**Pull request by szigetics**
As far as I know, when the following commands are being run from a MacOS arm64 machine, the build output will have the arm64 slice : 
```
> ./configure-mac_os_x.sh
> make all
```
, but : in many cases people need to be able to cross-compile from x86_64 to arm64. I added support for this workflow in this PR.

**Changes** : 
1.: Added MacOS Apple Silicon (arm64) build support
To build for arm64, the `--arm64` flag has to be passed to `configure-mac_os_x.sh` so the command looks like this : 
`./configure-mac_os_x.sh --arm64` 

I don't have access to Apple Silicon hardware at the moment. (My machine with M1 should arrive soon...)
This means that I couldn't run the unit tests yet on an actual M1 machine.
At least here is a quick confirmation that all the binaries and libraries are built for arm64 : 
```
> ./configure-mac_os_x.sh --arm64

*******************************************
PcapPlusPlus Mac OS X configuration script 
*******************************************

Number of arguments: 1

PcapPlusPlus configuration is complete. Files created (or modified): mk/platform.mk, mk/PcapPlusPlus.mk, mk/install.sh, mk/uninstall.sh
> make all
...
> cd Dist
> find ./ -name "*.a" -exec lipo -info {} \;
Non-fat file: .//libPcap++.a is architecture: arm64
Non-fat file: .//libCommon++.a is architecture: arm64
Non-fat file: .//libPacket++.a is architecture: arm64
> find examples -type f -exec lipo -info {} \;
Non-fat file: examples/IPFragUtil is architecture: arm64
Non-fat file: examples/Arping is architecture: arm64
Non-fat file: examples/HttpAnalyzer is architecture: arm64
Non-fat file: examples/IPDefragUtil is architecture: arm64
Non-fat file: examples/IcmpFileTransfer-catcher is architecture: arm64
Non-fat file: examples/SSLAnalyzer is architecture: arm64
Non-fat file: examples/PcapSplitter is architecture: arm64
Non-fat file: examples/IcmpFileTransfer-pitcher is architecture: arm64
Non-fat file: examples/DnsSpoofing is architecture: arm64
Non-fat file: examples/PcapSearch is architecture: arm64
Non-fat file: examples/DNSResolver is architecture: arm64
Non-fat file: examples/PcapPrinter is architecture: arm64
Non-fat file: examples/ArpSpoofing is architecture: arm64
Non-fat file: examples/TcpReassembly is architecture: arm64
Non-fat file: examples/TLSFingerprinting is architecture: arm64
```
Maybe until we won't have access to an Apple Silicon machine, the arm64 build support should be tagged with experimental label in the documentation...?

Relevant GitHub issue : https://github.com/seladb/PcapPlusPlus/issues/596 .

TODO : 
- [x] CI support
- [x] Use GLOBAL_FLAGS
- [x] Revert .cirrus.yml change
- [x] Apply review suggestions

### CODE REVIEW

File: mk/PcapPlusPlus.mk.macosx.arm64 [hunk 1]
```diff
@@ -0,0 +1,11 @@
+### MAC OS X ###
+
+# includes
+PCAPPP_INCLUDES += -I$(MACOS_SDK_HOME)/usr/include/netinet
+
+# libs
+PCAPPP_LIBS += -lpcap -lpthread -framework SystemConfiguration -framework CoreFoundation
```
Comment (szigetics) [inline mk/PcapPlusPlus.mk.macosx.arm64:7]: I am not really happy with the code duplication here. 
( Lines copied from here : https://github.com/seladb/PcapPlusPlus/blob/master/mk/PcapPlusPlus.mk.macosx#L3-L7 )
Could we / Should we include `PcapPlusPlus.mk.macosx` instead?

File: configure-mac_os_x.sh [hunk 2]
```diff
@@ -126,13 +133,20 @@ cp -f mk/platform.mk.macosx $PLATFORM_MK
...
 if [[ $MACOS_MINOR_VERSION -ge 14 ]]; then
    echo -e "\n# setting SDK home for MacOS version >= 10.14" >> $PCAPPLUSPLUS_MK
-   echo -e "MACOS_SDK_HOME := /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk\n" >> $PCAPPLUSPLUS_MK
+   echo -e "MACOS_SDK_HOME := $MACOS_SDK_HOME\n" >> $PCAPPLUSPLUS_MK
+fi
+
+PCAPPLUSPLUS_MK_MACOSX="PcapPlusPlus.mk.macosx"
+if [ -n "$BUILD_FOR_ARM64" ]; then
+   PCAPPLUSPLUS_MK_MACOSX="PcapPlusPlus.mk.macosx.arm64"
+   echo -e "MACOS_APPLE_SILICON_ARM64_BUILD_FLAGS := -arch arm64 -target arm64-apple-macos11 -isysroot $MACOS_SDK_HOME" >> $PLATFORM_MK
```
Comment (szigetics) [inline configure-mac_os_x.sh:146]: 👀  I am looking for a suggestion here : 

I am not really happy with the code duplication here.
( Lines copied from : https://github.com/seladb/PcapPlusPlus/blob/bd26f31b8229f0931542a6cba10436440d733cd9/mk/PcapPlusPlus.mk.macosx.arm64#L10 )

What would be the best way to clean it up?

**Comment by szigetics** [comment]
Should we update the travis.yml file as well ( https://github.com/seladb/PcapPlusPlus/blob/dev/.travis.yml )?

I could think of : 

1. Dropping xcode10.1 build (Xcode 10.x is already 2 or 3 years old)
2. Adding Xcode 12.3 support ( seems like we will have to use Xcode12.3 : https://travis-ci.community/t/osx-image-xcode12-2-does-not-come-with-macos-11-sdk-no-way-to-compile-for-arm/10611/8 ) : 

- Building for x86_64 using Xcode12.3, and running the unit tests
- Building for arm64 using Xcode12.3, skipping running the unit tests (because we don't have an arm64 machine to run the tests on), checking only that the built libraries and binaries are containing the `arm64` slice (and the `arm64` slice only)

**Comment by szigetics** [comment]
I see that the documentation about building is not part of the repo : 
https://pcapplusplus.github.io/docs/install/build-source/macos

It would be nice not to forget about updating these docs as well.

File: configure-mac_os_x.sh (hunk 2 above)
Comment (seladb) [reply configure-mac_os_x.sh:146]: I'm currently adding support for Android and have the same issue:
https://github.com/seladb/PcapPlusPlus/blob/android-support/3rdParty/LightPcapNg/Makefile#L14-L19
https://github.com/seladb/PcapPlusPlus/blob/android-support/3rdParty/MemPlumber/Makefile#L9-L14
https://github.com/seladb/PcapPlusPlus/blob/android-support/3rdParty/hash-library/Makefile#L14-L19

I'm thinking that maybe we can add a new makefile where we can put these flags, and then include it in the 3rd-party Makefiles.
For example: we can add `3rdParty.mk.macosx.arm64` that includes these build flags and then pour its content to `3rdParty.mk` which we can include in the 3rd-party Makefiles.
Once you do this for Apple Silicon I can use the method for Android.
Please let me know what you think.

**Review by seladb — COMMENTED** [review]
Since this is cross-compiling I'd suggest adding this to the CI, maybe we can add it to the `macos` build: 
https://github.com/seladb/PcapPlusPlus/blob/c08390d0178093b310d9de31a82796b8f5083c30/.github/workflows/build_and_test.yml#L121

Also, do you expect to get your M1 machine soon?
Unfortunately no CI supports M1 as of now so there is no way to test it continuously, but if you can test it at least once on your machine that would be great.

**Comment by seladb** [comment]
I thought about it some more and came to the conclusion it'd make more sense to add `GLOBAL_FLAGS` to `mk/platform.mk.*`.
This is what I did for Android support:
- Added `GLOBAL_FLAGS`: https://github.com/seladb/PcapPlusPlus/blob/90be4eece33033684da95341292a558a34f60902/mk/platform.mk.android#L27
- Then use it in all of the 3rd-party libs:
  https://github.com/seladb/PcapPlusPlus/blob/90be4eece33033684da95341292a558a34f60902/3rdParty/LightPcapNg/Makefile#L18
  https://github.com/seladb/PcapPlusPlus/blob/90be4eece33033684da95341292a558a34f60902/3rdParty/hash-library/Makefile#L16
  https://github.com/seladb/PcapPlusPlus/blob/90be4eece33033684da95341292a558a34f60902/3rdParty/MemPlumber/Makefile#L11

I think you can follow the same solution, please let me know what you think.

**Comment by szigetics** [comment]
> I thought about it some more and came to the conclusion it'd make more sense to add `GLOBAL_FLAGS` to `mk/platform.mk.*`.
> This is what I did for Android support:
> 
> * Added `GLOBAL_FLAGS`: https://github.com/seladb/PcapPlusPlus/blob/90be4eece33033684da95341292a558a34f60902/mk/platform.mk.android#L27
> * Then use it in all of the 3rd-party libs:
>   https://github.com/seladb/PcapPlusPlus/blob/90be4eece33033684da95341292a558a34f60902/3rdParty/LightPcapNg/Makefile#L18
>   
>   https://github.com/seladb/PcapPlusPlus/blob/90be4eece33033684da95341292a558a34f60902/3rdParty/hash-library/Makefile#L16
>   
>   https://github.com/seladb/PcapPlusPlus/blob/90be4eece33033684da95341292a558a34f60902/3rdParty/MemPlumber/Makefile#L11
> 
> I think you can follow the same solution, please let me know what you think.

Looks good. I am going to follow the same approach.

**Comment by szigetics** [comment]
> Since this is cross-compiling I'd suggest adding this to the CI, maybe we can add it to the `macos` build:
> https://github.com/seladb/PcapPlusPlus/blob/c08390d0178093b310d9de31a82796b8f5083c30/.github/workflows/build_and_test.yml#L121
> 
> Also, do you expect to get your M1 machine soon?
> Unfortunately no CI supports M1 as of now so there is no way to test it continuously, but if you can test it at least once on your machine that would be great.

I just checked and it seems that the Github Workflow was manually disabled : 
Screenshot 2021-03-29 at 08 18 11
I see that only Cirrus CI is running for the PRs so for now I am adding the arm64 cross-compiling build to cirrus.
Please let me know if we should really add it to Github actions instead / as well!

I will receive my M1 machine this Wednesday. I am going to test the build this week.

**Comment by seladb** [comment]
@szigetics the `macOS` workflow is not in use anymore. The only active workflow is `Build and test` which contains all of the platforms. You can find the `macOS` build here:
https://github.com/seladb/PcapPlusPlus/blob/c08390d0178093b310d9de31a82796b8f5083c30/.github/workflows/build_and_test.yml#L121

**Comment by seladb** [comment]
I've just merge Android support into `dev` branch, which should make it easier for you to integrate. Please let me know if any issues

**Comment by szigetics** [comment]
> @szigetics the `macOS` workflow is not in use anymore. The only active workflow is `Build and test` which contains all of the platforms. You can find the `macOS` build here:
> https://github.com/seladb/PcapPlusPlus/blob/c08390d0178093b310d9de31a82796b8f5083c30/.github/workflows/build_and_test.yml#L121

Isn't it an issue that "This workflow was disabled manually." ? ( Ref.: https://github.com/seladb/PcapPlusPlus/actions/workflows/macos.yml )
So these Github Workflow CI pipelines aren't currently running for the PRs.

**Comment by seladb** [comment]
> Isn't it an issue that "This workflow was disabled manually." ? ( Ref.: https://github.com/seladb/PcapPlusPlus/actions/workflows/macos.yml )
> So these Github Workflow CI pipelines aren't currently running for the PRs.

These GitHub Workflow pipelines shouldn't be running. I've used them before but then I consolidated everything to one pipeline which is `build_and_test` and this is the only one that should be running.

**Comment by szigetics** [comment]
> > Isn't it an issue that "This workflow was disabled manually." ? ( Ref.: https://github.com/seladb/PcapPlusPlus/actions/workflows/macos.yml )
> > So these Github Workflow CI pipelines aren't currently running for the PRs.
> 
> These GitHub Workflow pipelines shouldn't be running. I've used them before but then I consolidated everything to one pipeline which is `build_and_test` and this is the only one that should be running.

Do you have this enabled in the repo settings?
88571581-de3d5300-d00b-11ea-8d8c-027e951d63e1

I don't see the build_and_test workflow running for my PR (which is using a fork).

**Comment by szigetics** [comment]
@seladb : 
I am done for now. Looking for your feedback! :)

File: .cirrus.yml [hunk 3]
```diff
@@ -68,6 +68,18 @@ task:
...
+  name: Test Apple Silicon (arm64) MacOS
+  osx_instance:
+    image: catalina-xcode
+  configure_script:
+    - ./configure-mac_os_x.sh --arm64
+  build_script:
+    - make all
+  check_for_arm64_slice_script:
+    - cd Dist && ALL_EXECUTABLES_WITHOUT_SYMLINKS=($(find . -perm +111 -type f)) && for EXECUTABLE in ${ALL_EXECUTABLES_WITHOUT_SYMLINKS[@]}; do echo "Checking for arm64 in $EXECUTABLE"; CHECK_ARCH=$(lipo $EXECUTABLE -verify_arch arm64 2>&1); RETVAL=$?; if [ $RETVAL -eq 1 ]; then echo "Couldn't find arm64 slice in $EXECUTABLE"; exit 1; else echo "Found arm64 slice in $EXECUTABLE"; fi; done
+    - cd ../Dist && ALL_LIBRARIES_WITHOUT_SYMLINKS=($(find . -name "*.a" -type f)) &&  for LIBRARY in ${ALL_LIBRARIES_WITHOUT_SYMLINKS[@]}; do echo "Checking for arm64 in $LIBRARY"; CHECK_ARCH=$(lipo $LIBRARY -verify_arch arm64 2>&1); RETVAL=$?; if [ $RETVAL -eq 1 ]; then echo "Couldn't find arm64 slice in $LIBRARY"; exit 1; else echo "Found arm64 slice in $LIBRARY"; fi; done
```
Comment (seladb) [inline .cirrus.yml:81]: I think we can remove that part? 🤔

File: mk/PcapPlusPlus.mk.macosx.arm64 [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+# Apple Silicon (arm64) specific build flags
+CXXFLAGS += -arch arm64 -target arm64-apple-macos11 -isysroot $(MACOS_SDK_HOME)
+PCAPPP_BUILD_FLAGS += $(CXXFLAGS)
```
Comment (seladb) [inline mk/PcapPlusPlus.mk.macosx.arm64:11]: I'd suggest the following:
- Keep only this part in `PcapPlusPlus.mk.macosx.arm64`
- In `configure-mac_os_x.sh`  you can do something like this:
  ```shell
  cat mk/PcapPlusPlus.mk.macosx >> $PCAPPLUSPLUS_MK
  if [ -n "$BUILD_FOR_ARM64" ]; then
    cat mk/PcapPlusPlus.mk.macosx.arm64 >> $PCAPPLUSPLUS_MK
  fi
  ```

File: configure-mac_os_x.sh [hunk 5]
```diff
@@ -126,13 +133,20 @@ cp -f mk/platform.mk.macosx $PLATFORM_MK
...
 if [[ $MACOS_MINOR_VERSION -ge 14 ]]; then
    echo -e "\n# setting SDK home for MacOS version >= 10.14" >> $PCAPPLUSPLUS_MK
-   echo -e "MACOS_SDK_HOME := /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk\n" >> $PCAPPLUSPLUS_MK
+   echo -e "MACOS_SDK_HOME := $MACOS_SDK_HOME\n" >> $PCAPPLUSPLUS_MK
+fi
+
+PCAPPLUSPLUS_MK_MACOSX="PcapPlusPlus.mk.macosx"
+if [ -n "$BUILD_FOR_ARM64" ]; then
+   PCAPPLUSPLUS_MK_MACOSX="PcapPlusPlus.mk.macosx.arm64"
+   echo -e "GLOBAL_FLAGS := -arch arm64 -target arm64-apple-macos11 -isysroot $MACOS_SDK_HOME" >> $PLATFORM_MK
```
Comment (seladb) [inline configure-mac_os_x.sh:146]: I'd suggest moving this line to a new file `platform.mk.macosx.arm64` and then do:

```shell
if [ -n "$BUILD_FOR_ARM64" ]; then
  cat mk/platform.mk.macosx.arm64 >> $PLATFORM_MK
fi
```

File: .github/workflows/build_and_test.yml [hunk 6]
```diff
@@ -167,6 +167,38 @@ jobs:
         cd Examples/Tutorials/Tutorial-HelloWorld
         make -f Makefile.non_windows all
         ./Tutorial-HelloWorld
+        
+  macos-apple-silicon-arm64:
```
Comment (seladb) [inline .github/workflows/build_and_test.yml:171]: I've updated `build_and_test.yml` and it now triggers on PR. You can sync this branch to PcapPlusPlus `dev` and it should trigger for the PR

Comment (szigetics) [reply .github/workflows/build_and_test.yml:171]: Thank you! Done.

File: .cirrus.yml (hunk 3 above)
Comment (szigetics) [reply .cirrus.yml:81]: @seladb  : Do you mean to remove the whole `.cirrus.yml` update or to remove only the arm64 slice check?

Comment (seladb) [reply .cirrus.yml:81]: I think we can remove the `.cirrus.yml` task if we have a similar task that runs in GitHub Actions, right?

Comment (szigetics) [reply .cirrus.yml:81]: Yes, sure.

**Comment by szigetics** [comment]
@seladb :
I am done for now. Looking for your feedback! 🙂

**Comment by szigetics** [comment]
@seladb : 
I ran all the tests on my M1 machine. All of the tests that I could run are passing. Here are the test results : 
```
Tests % cd Packet++Test 
Packet++Test % Bin/Packet++Test 
PcapPlusPlus version: v20.08+ (non-official release)
Built: Apr  3 2021 14:51:37
Built from: Git branch 'macos-apple-silicon-arm64-support', commit 'a81b998dc96bb0485ae50c3d244bcbccf823aa13'
Start running tests...

EthPacketCreation             : PASSED
EthPacketPointerCreation      : PASSED
EthAndArpPacketParsing        : PASSED
ArpPacketCreation             : PASSED
EthDot3LayerParsingTest       : PASSED
EthDot3LayerCreateEditTest    : PASSED
VlanParseAndCreation          : PASSED
MplsLayerTest                 : PASSED
VxlanParsingAndCreationTest   : PASSED
IPv4PacketCreation            : PASSED
IPv4PacketParsing             : PASSED
IPv4FragmentationTest         : PASSED
IPv4OptionsParsingTest        : PASSED
IPv4OptionsEditTest           : PASSED
IPv4UdpChecksum               : PASSED
IPv6UdpPacketParseAndCreate   : PASSED
IPv6FragmentationTest         : PASSED
IPv6ExtensionsTest            : PASSED
TcpPacketNoOptionsParsing     : PASSED
TcpPacketWithOptionsParsing   : PASSED
TcpPacketWithOptionsParsing2  : PASSED
TcpPacketCreation             : PASSED
TcpPacketCreation2            : PASSED
TcpMalformedPacketParsing     : PASSED
PacketUtilsHash5TupleUdp      : PASSED
PacketUtilsHash5TupleTcp      : PASSED
PacketUtilsHash5TupleIPv6     : PASSED
InsertDataToPacket            : PASSED
InsertVlanToPacket            : PASSED
RemoveLayerTest               : PASSED
CopyLayerAndPacketTest        : PASSED
PacketLayerLookupTest         : PASSED
RawPacketTimeStampSetterTest  : PASSED
ParsePartialPacketTest        : PASSED
PacketTrailerTest             : PASSED
ResizeLayerTest               : PASSED
HttpRequestLayerParsingTest   : PASSED
HttpRequestLayerCreationTest  : PASSED
HttpRequestLayerEditTest      : PASSED
HttpResponseLayerParsingTest  : PASSED
HttpResponseLayerCreationTest : PASSED
HttpResponseLayerEditTest     : PASSED
HttpMalformedResponseTest     : PASSED
PPPoESessionLayerParsingTest  : PASSED
PPPoESessionLayerCreationTest : PASSED
PPPoEDiscoveryLayerParsingTest: PASSED
PPPoEDiscoveryLayerCreateTest : PASSED
DnsLayerParsingTest           : PASSED
DnsLayerQueryCreationTest     : PASSED
DnsLayerResourceCreationTest  : PASSED
DnsLayerEditTest              : PASSED
DnsLayerRemoveResourceTest    : PASSED
DnsOverTcpParsingTest         : PASSED
DnsOverTcpCreationTest        : PASSED
IcmpParsingTest               : PASSED
IcmpCreationTest              : PASSED
IcmpEditTest                  : PASSED
GreParsingTest                : PASSED
GreCreationTest               : PASSED
GreEditTest                   : PASSED
SSLClientHelloParsingTest     : PASSED
SSLAppDataParsingTest         : PASSED
SSLAlertParsingTest           : PASSED
SSLMultipleRecordParsingTest  : PASSED
SSLMultipleRecordParsing2Test : PASSED
SSLMultipleRecordParsing3Test : PASSED
SSLMultipleRecordParsing4Test : PASSED
SSLPartialCertificateParseTest: PASSED
SSLNewSessionTicketParseTest  : PASSED
SSLMalformedPacketParsing     : PASSED
TLS1_3ParsingTest             : PASSED
TLSCipherSuiteTest            : PASSED
ClientHelloTLSFingerprintTest : PASSED
ServerHelloTLSFingerprintTest : PASSED
SllPacketParsingTest          : PASSED
SllPacketCreationTest         : PASSED
NullLoopbackTest              : PASSED
DhcpParsingTest               : PASSED
DhcpCreationTest              : PASSED
DhcpEditTest                  : PASSED
IgmpParsingTest               : PASSED
IgmpCreateAndEditTest         : PASSED
Igmpv3ParsingTest             : PASSED
Igmpv3QueryCreateAndEditTest  : PASSED
Igmpv3ReportCreateAndEditTest : PASSED
SipRequestLayerParsingTest    : PASSED
SipRequestLayerCreationTest   : PASSED
SipRequestLayerEditTest       : PASSED
SipResponseLayerParsingTest   : PASSED
SipResponseLayerCreationTest  : PASSED
SipResponseLayerEditTest      : PASSED
SdpLayerParsingTest           : PASSED
SdpLayerCreationTest          : PASSED
SdpLayerEditTest              : PASSED
RadiusLayerParsingTest        : PASSED
RadiusLayerCreationTest       : PASSED
RadiusLayerEditTest           : PASSED
GtpLayerParsingTest           : PASSED
GtpLayerCreationTest          : PASSED
GtpLayerEditTest              : PASSED
BgpLayerParsingTest           : PASSED
BgpLayerCreationTest          : PASSED
BgpLayerEditTest              : PASSED
SSHParsingTest                : PASSED
SSHMalformedParsingTest       : PASSED
IPSecParsingTest              : PASSED

ALL TESTS PASSED!!
Test cases: 106, Passed: 106, Failed: 0, Skipped: 0
```
+
```
Tests % cd Pcap++Test 
Pcap++Test % Bin/Pcap++Test -n
Running only tests that don't require network connection
PcapPlusPlus version: v20.08+ (non-official release)
Built: Apr  3 2021 14:52:01
Git info: Git branch 'macos-apple-silicon-arm64-support', commit 'a81b998dc96bb0485ae50c3d244bcbccf823aa13'
Using ip: 
Debug mode: off
Start running tests...

TestIPAddress                 : PASSED
TestMacAddress                : PASSED
TestLRUList                   : PASSED
TestGeneralUtils              : PASSED
TestPcapFileReadWrite         : PASSED
TestPcapSllFileReadWrite      : PASSED
TestPcapRawIPFileReadWrite    : PASSED
TestPcapFileAppend            : PASSED
TestPcapNgFileReadWrite       : PASSED
TestPcapNgFileReadWriteAdv    : PASSED
TestPcapFileReadLinkTypeIPv6  : PASSED
TestPcapFileReadLinkTypeIPv4  : PASSED
TestPcapLiveDeviceList        : PASSED
TestPcapLiveDeviceNoNetworking: PASSED
TestPcapFilters_MatchStatic   : PASSED
TestPcapFilters_General_BPFStr: PASSED
TestPcapFiltersOffline        : PASSED
TestPcapFilters_LinkLayer     : PASSED
TestHttpRequestParsing        : PASSED
TestHttpResponseParsing       : PASSED
TestPrintPacketAndLayers      : PASSED
TestDnsParsing                : PASSED
TestTcpReassemblySanity       : PASSED
TestTcpReassemblyRetran       : PASSED
TestTcpReassemblyMissingData  : PASSED
TestTcpReassemblyOutOfOrder   : PASSED
TestTcpReassemblyWithFIN_RST  : PASSED
TestTcpReassemblyMalformedPkts: PASSED
TestTcpReassemblyMultipleConns: PASSED
TestTcpReassemblyIPv6         : PASSED
TestTcpReassemblyIPv6MultConns: PASSED
TestTcpReassemblyIPv6_OOO     : PASSED
TestTcpReassemblyCleanup      : PASSED
TestTcpReassemblyMaxOOOFrags  : PASSED
TestTcpReassemblyMaxSeq       : PASSED
TestIPFragmentationSanity     : PASSED
TestIPFragOutOfOrder          : PASSED
TestIPFragPartialData         : PASSED
TestIPFragMultipleFrags       : PASSED
TestIPFragMapOverflow         : PASSED
TestIPFragRemove              : PASSED

ALL TESTS PASSED!!
Test cases: 68, Passed: 41, Failed: 0, Skipped: 27
```
+
I couldn't run the `pytest` tests though : 
```
Tests % cd ExamplesTest
ExamplesTest % pip3 install requirements.txt
WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.
Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
Defaulting to user installation because normal site-packages is not writeable
ERROR: Could not find a version that satisfies the requirement requirements.txt
ERROR: No matching distribution found for requirements.txt
```
+
```
PcapPlusPlus % make install
mkdir: /usr/local/lib: Permission denied
make: *** [install] Error 1
PcapPlusPlus % sudo make install
Installation complete!
PcapPlusPlus % cd Examples/Tutorials/Tutorial-HelloWorld 
Tutorial-HelloWorld % make -f Makefile.non_windows all
g++ -fPIC -O2 -g -Wall -O2 -g -Wall -arch arm64 -target arm64-apple-macos11 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include/pcapplusplus -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/netinet -c -o main.o main.cpp
g++  -static-libstdc++ -o Tutorial-HelloWorld main.o -lPcap++ -lPacket++ -lCommon++ -lpcap -lpthread -framework SystemConfiguration -framework CoreFoundation
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
Tutorial-HelloWorld % ./Tutorial-HelloWorld 
Source IP is '10.0.0.138'; Dest IP is '10.0.0.1'
Tutorial-HelloWorld % echo $?
0
```

**Comment by seladb** [comment]
Thank you so much @szigetics for all your efforts and hard work!
The PR is ready to be merged!
Thank you also for testing it on your M1 machine, it's very encouraging to see that all of the tests pass. 

One more thing that would be really great trying is to run `Pcap++Test` with network connection, meaning:

```shell
Tests % cd Pcap++Test 
Pcap++Test % Bin/Pcap++Test -i [YOUR_IP_ADDRESS]
```

Assuming that the libpcap version that you have is compatible with M1 (which I guess it should) I think all the tests should pass, but it's worth making sure. While you run the tests, please keep network traffic flowing (for example: have a YouTube video running or browse to some web-sites). Thanks!

**Comment by szigetics** [comment]
> Thank you so much @szigetics for all your efforts and hard work!
> The PR is ready to be merged!
> Thank you also for testing it on your M1 machine, it's very encouraging to see that all of the tests pass.
> 
> One more thing that would be really great trying is to run `Pcap++Test` with network connection, meaning:
> 
> ```shell
> Tests % cd Pcap++Test 
> Pcap++Test % Bin/Pcap++Test -i [YOUR_IP_ADDRESS]
> ```
> 
> Assuming that the libpcap version that you have is compatible with M1 (which I guess it should) I think all the tests should pass, but it's worth making sure. While you run the tests, please keep network traffic flowing (for example: have a YouTube video running or browse to some web-sites). Thanks!

Thank you for your hard work and for the quick review! 😊

I ran these additional tests. These ones have passed as well.
I had to call it with sudo otherwise it gave some errors like this : 
```
(cannot open BPF device) /dev/bpf0: Permission denied
(cannot open BPF device) /dev/bpf0: Permission denied
TestGetMacAddress             : FAILED (Tests/IpMacTests.cpp:250). assert true failed: liveDev->open()
```

Here is the output : 
```
% sudo Bin/Pcap++Test -i 172.20.10.2    
PcapPlusPlus version: v20.08+ (non-official release)
Built: Apr  3 2021 14:52:01
Git info: Git branch 'macos-apple-silicon-arm64-support', commit 'a81b998dc96bb0485ae50c3d244bcbccf823aa13'
Using ip: 172.20.10.2
Debug mode: off
Start running tests...

TestIPAddress                 : PASSED
TestMacAddress                : PASSED
TestLRUList                   : PASSED
TestGeneralUtils              : PASSED
TestGetMacAddress             : PASSED
TestPcapFileReadWrite         : PASSED
TestPcapSllFileReadWrite      : PASSED
TestPcapRawIPFileReadWrite    : PASSED
TestPcapFileAppend            : PASSED
TestPcapNgFileReadWrite       : PASSED
TestPcapNgFileReadWriteAdv    : PASSED
TestPcapFileReadLinkTypeIPv6  : PASSED
TestPcapFileReadLinkTypeIPv4  : PASSED
TestPcapLiveDeviceList        : PASSED
TestPcapLiveDeviceListSearch  : PASSED
TestPcapLiveDevice            : PASSED
TestPcapLiveDeviceNoNetworking: PASSED
TestPcapLiveDeviceStatsMode   : PASSED
TestPcapLiveDeviceBlockingMode: PASSED
TestPcapLiveDeviceSpecialCfg  : PASSED
TestWinPcapLiveDevice         : PASSED
TestSendPacket                : PASSED
TestSendPackets               : PASSED
TestRemoteCapture             : PASSED
TestPcapFilters_MatchStatic   : PASSED
TestPcapFiltersLive           : PASSED
TestPcapFilters_General_BPFStr: PASSED
TestPcapFiltersOffline        : PASSED
TestPcapFilters_LinkLayer     : PASSED
TestHttpRequestParsing        : PASSED
TestHttpResponseParsing       : PASSED
TestPrintPacketAndLayers      : PASSED
TestDnsParsing                : PASSED
TestTcpReassemblySanity       : PASSED
TestTcpReassemblyRetran       : PASSED
TestTcpReassemblyMissingData  : PASSED
TestTcpReassemblyOutOfOrder   : PASSED
TestTcpReassemblyWithFIN_RST  : PASSED
TestTcpReassemblyMalformedPkts: PASSED
TestTcpReassemblyMultipleConns: PASSED
TestTcpReassemblyIPv6         : PASSED
TestTcpReassemblyIPv6MultConns: PASSED
TestTcpReassemblyIPv6_OOO     : PASSED
TestTcpReassemblyCleanup      : PASSED
TestTcpReassemblyMaxOOOFrags  : PASSED
TestTcpReassemblyMaxSeq       : PASSED
TestIPFragmentationSanity     : PASSED
TestIPFragOutOfOrder          : PASSED
TestIPFragPartialData         : PASSED
TestIPFragMultipleFrags       : PASSED
TestIPFragMapOverflow         : PASSED
TestIPFragRemove              : PASSED
TestRawSockets                : PASSED

ALL TESTS PASSED!!
Test cases: 68, Passed: 53, Failed: 0, Skipped: 15
```

**Comment by seladb** [comment]
That is great, thank you! I'll merge the PR to `master` later and also update the documentation to reflect M1 support

**Pull request merged**